README.md
<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version-last-release/hySpc.read.spe)](https://cran.r-project.org/package=hySpc.read.spe)
[![R-CMD-check](https://github.com/r-hyperspec/hySpc.read.spe/workflows/R-CMD-check/badge.svg)](https://github.com/r-hyperspec/hySpc.read.spe/actions)
![Website (pkgdown)](https://github.com/r-hyperspec/hySpc.read.spe/workflows/Website%20(pkgdown)/badge.svg)
[![Codecov](https://codecov.io/gh/r-hyperspec/hySpc.read.spe/branch/develop/graph/badge.svg)](https://codecov.io/gh/r-hyperspec/hySpc.read.spe?branch=develop)
[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
<!--[![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/hySpc.read.spe)](https://cran.r-project.org/package=hySpc.read.spe)-->
<!--[![metacran downloads](https://cranlogs.r-pkg.org/badges/hySpc.read.spe)](https://cran.r-project.org/package=hySpc.read.spe)-->
<!-- badges: end -->



# R Package **hySpc.read.spe**

[**R**](https://www.r-project.org/) package **hySpc.read.spe** is a member of the [**`r-hyperspec`**](https://r-hyperspec.github.io/) packages family, which imports a number of file formats produced by Princeton Instruments spectrometers.
    The imported data are returned as hyperSpec objects.

<!-- ---------------------------------------------------------------------- -->

## Documentation

There are two versions of **hySpc.read.spe** online documentation:

a. for the [released version](https://r-hyperspec.github.io/hySpc.read.spe/) of package,  
b. for the [development version](https://r-hyperspec.github.io/hySpc.read.spe/dev/) of package.

The documentation of the other **`r-hyperspec`** family packages can be found at [r-hyperspec.github.io](https://r-hyperspec.github.io/).

<!-- ---------------------------------------------------------------------- -->

## Installation

### Install from CRAN

You can install the released version of **hySpc.read.spe** from [CRAN](https://cran.r-project.org/package=hySpc.read.spe) with:

```r
install.packages("hySpc.read.spe")
```


### Install from GitHub

You can install the development version of the package from [GitHub](https://github.com/r-hyperspec/hySpc.read.spe):

```r
if (!require(remotes)) {install.packages("remotes")}
remotes::install_github("r-hyperspec/hySpc.read.spe")
```

**NOTE 1:**
Usually, "Windows" users need to download, install and properly configure **Rtools** (see [these instructions](https://cran.r-project.org/bin/windows/Rtools/)) to make the code above work.

**NOTE 2:**
This method will **not** install package's documentation (help pages and vignettes) into your computer.
So you can either use the [online documentation](https://r-hyperspec.github.io/) or build the package from source (see the next section).


### Install from Source

1. From the **hySpc.read.spe**'s GitHub [repository](https://github.com/r-hyperspec/hySpc.read.spe):
    - If you use Git, `git clone` the branch of interest.
      You may need to fork it before cloning.
    - Or just chose the branch of interest (1 in Figure below), download a ZIP archive with the code (2, 3) and unzip it on your computer.  
![image](https://user-images.githubusercontent.com/12725868/89338263-ffa1dd00-d6a4-11ea-94c2-fa36ee026691.png)

2. Open the downloaded directory in RStudio (preferably, as an RStudio project).
    - The code below works correctly only if your current working directory coincides with the root of the repository, i.e., if it is in the directory that contains file `README.md`.
    - If you open RStudio project correctly (e.g., by clicking `project.Rproj` icon ![image](https://user-images.githubusercontent.com/12725868/89340903-26621280-d6a9-11ea-8299-0ec5e9cf7e3e.png) in the directory), then the working directory is set correctly by default.

3. In RStudio 'Console' window, run the code (provided below) to:
    a. Install packages **remotes** and **devtools**.
    b. Install **hySpc.read.spe**'s dependencies.
    c. Create **hySpc.read.spe**'s documentation.
    d. Install package **hySpc.read.spe**.

```r
# Do not abort installation even if some packages are not available
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS = "true")

# Install packages remotes and devtools
install.packages(c("remotes", "devtools"))

# Install hySpc.read.spe's dependencies
remotes::install_deps(dependencies = TRUE)

# Create hySpc.read.spe's documentation
devtools::document()

# Install package hySpc.read.spe
devtools::install(build_vignettes = TRUE)
```

**NOTE 1:**
Usually, "Windows" users need to download, install and properly configure **Rtools** (see [these instructions](https://cran.r-project.org/bin/windows/Rtools/)) to make the code above work.


NEWS.md
# hySpc.read.spe 0.0.x 2021-xx-xx

## Acknowledgements
The `hyperSpec` team gratefully acknowledges support from the Google Summer of Code program, which sponsored student Sang Truong during summer 2021.
Sang and the team carried out a significant overhaul of `hyperSpec` which led to this release.

## User-Visible Changes

* Moved functions from **hyperSpec** to **hySpc.read.spe**:
  - `hyperSpec::read.spe()` -> `hySpc.read.spe::read_spe()`
  - `hyperSpec::spe.showcalpoints()` -> `hySpc.read.spe::spe_plot_calibration_points()`
  - `hyperSpec::.read.spe.xml()` -> `hySpc.read.spe::read_spe_xml()`
  - `hyperSpec::.read.spe.xml_string()` -> `hySpc.read.spe::read_spe_xml_string()`
  - `hyperSpec::.read.spe.header()` -> `hySpc.read.spe::read_spe_header()`
  
* In `read_spe()` and `spe_plot_calibration_points()`, argument `xaxis` was renamed to `wl_units` and default value was changed from `"file`" to `NULL` (#20, #21).


## New Features

* When new versions of the data are pushed to this repo, the newly created `.tar.gz` is automatically deployed to `pkg-repo`.


tests/testthat.R
library(testthat)
library(hySpc.read.spe)

test_check("hySpc.read.spe")


tests/testthat/test_attached.R
# C. Beleites
# run all tests attached to objects
hySpc.testthat::unittest("hySpc.read.spe", standalone = FALSE)


R/wl_convert_units.R

#' Convert between Different Wavelength Units
#'
#' The following units can be converted into each other:
#' *nm*, \emph{\eqn{cm^{-1}}{inverse cm}}, *eV*, *THz* and *Raman shift*
#'
#' @param x data for conversion
#' @param from source unit
#' @param to destination unit
#' @param ref_wl laser wavelength (required for work with Raman shift)
#' @author R. Kiselev
#' @export
#'
#' @concept wavelengths
#'
#' @examples
#' wl_convert_units(3200, "Raman shift", "nm", ref_wl = 785.04)
#' wl_convert_units(785, "nm", "invcm")
wl_convert_units <- function(x, from, to, ref_wl = NULL) {
  src  <- .wl_fix_unit_name(from)
  dest <- .wl_fix_unit_name(to)

  if (src == dest) {
    return(x)
  }

  if ((src == "raman" | dest == "raman") & is.null(ref_wl)) {
    stop("Working with Raman shift requires knowledge of laser wavelength")
  }

  f <- paste0("wl_", src, "2", dest)
  f <- get(f)
  return(f(x, ref_wl))
}

wl_ev2freq     <- function(x, ...)    wl_nm2freq(wl_ev2nm(x))
wl_ev2invcm    <- function(x, ...)    q * x / (100 * h * c)
wl_ev2nm       <- function(x, ...)    1e9 * h * c / (q * x)
wl_ev2raman    <- function(x, ref_wl) 1e7 / ref_wl - x * q / (100 * h * c)
wl_freq2ev     <- function(x, ...)    wl_nm2ev(wl_freq2nm(x))
wl_freq2invcm  <- function(x, ...)    wl_nm2invcm(wl_freq2nm(x))
wl_freq2nm     <- function(x, ...)    1e-3 * c / x
wl_freq2raman  <- function(x, ref_wl) wl_nm2raman(wl_freq2nm(x), ref_wl)
wl_invcm2ev    <- function(x, ...)    100 * x * c * h / q
wl_invcm2freq  <- function(x, ...)    wl_nm2freq(wl_invcm2nm(x))
wl_invcm2nm    <- function(x, ...)    1e7 / x
wl_invcm2raman <- function(x, ref_wl) 1e7 / ref_wl - x
wl_nm2ev       <- function(x, ...)    1e9 * h * c / (q * x)
wl_nm2freq     <- function(x, ...)    1e-3 * c / x
wl_nm2invcm    <- function(x, ...)    1e7 / x
wl_nm2raman    <- function(x, ref_wl) 1e7 * (1 / ref_wl - 1 / x)
wl_raman2ev    <- function(x, ref_wl) 100 * h * c * (1e7 / ref_wl - x) / q
wl_raman2freq  <- function(x, ref_wl) wl_nm2freq(wl_raman2nm(x, ref_wl))
wl_raman2invcm <- function(x, ref_wl) 1e7 / ref_wl - x
wl_raman2nm    <- function(x, ref_wl) 1e7 / (1e7 / ref_wl - x)


# Bring the argument to a conventional name
# FIXME: This function should be exported when `read.spe()` function is moved to
#        another package.
# FIXME: This function should be documented.
#        Even if it used for internal purposes.
.wl_fix_unit_name <- function(unit) {
  unit <- gsub(" .*$", "", tolower(unit))
  if (unit %in% c("raman", "stokes", "rel", "rel.", "relative", "rel.cm-1", "rel.cm", "rel.1/cm", "raman shift")) {
    return("raman")
  }
  if (unit %in% c("invcm", "energy", "wavenumber", "cm-1", "inverted", "cm", "1/cm")) {
    return("invcm")
  }
  if (unit %in% c("nm", "nanometer", "wavelength")) {
    return("nm")
  }
  if (unit %in% c("ev", "electronvolt")) {
    return("ev")
  }
  if (unit %in% c("freq", "frequency", "thz", "terahertz")) {
    return("freq")
  }
  if (unit %in% c("pixel", "px", "sensor")) {
    return("px")
  }
  if (unit == "file") {
    return(unit)
  }
  stop(paste0("'", unit, "': Unknown unit type"))
}


# Some physical constants ----------------------------------------------------
# @concept constants
q <- 1.60217656535e-19 # elementary charge
h <- 6.6260695729e-34 # Planck's constant
c <- 299792458 # speed of light


# Unit tests -----------------------------------------------------------------

hySpc.testthat::test(wl) <- function() {

  context("get wl")

  test_that("wl() works", {
    # Data
    hy_obj <- new("hyperSpec", spc = matrix(1:100, nrow = 1), wavelength = 601:700)

    # Perform tests
    expect_silent(res <- wl(hy_obj))
    expect_true(is.numeric(res)) # Can be either integer or double
    expect_length(res, 100)
    expect_equal(res, 601:700)
  })


  context("set wl")

  test_that("`wl<-` works", {
    # Data
    hy_obj <- new("hyperSpec", spc = matrix(1:100, nrow = 1), wavelength = 601:700)

    # Set new wavelengths
    expect_silent(wl(hy_obj) <- (1:nwl(hy_obj)) + 1000)
    expect_true(is.numeric(hy_obj@wavelength))
    expect_equal(hy_obj@wavelength, 1001:1100)

    # Set new wavelengths and label
    expect_equal(labels(hy_obj, ".wavelength"), ".wavelength")
    expect_silent(
      wl(hy_obj) <- list(wl = 101:200, label = "new label")
    )
    expect_equal(hy_obj@wavelength, 101:200)
    expect_equal(labels(hy_obj, ".wavelength"), "new label")
  })

}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hySpc.testthat::test(.wl_fix_unit_name) <- function() {

  context(".wl_fix_unit_name")

  test_that(".wl_fix_unit_name() works", {

    expect_equal(.wl_fix_unit_name("raman"), "raman")
    expect_equal(.wl_fix_unit_name("invcm"), "invcm")
    expect_equal(.wl_fix_unit_name("nm"),    "nm")
    expect_equal(.wl_fix_unit_name("ev"),    "ev")
    expect_equal(.wl_fix_unit_name("freq"),  "freq")
    expect_equal(.wl_fix_unit_name("px"),    "px")
    expect_equal(.wl_fix_unit_name("file"),  "file")
    expect_error(.wl_fix_unit_name("ddd"),   "Unknown unit type")

  })

  # TODO (tests): add more specific tests.
}

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hySpc.testthat::test(wl_convert_units) <- function() {

  context("wl_convert_units")

  test_that("wl_convert_units() throws error", {
    expect_error(wl_convert_units())

    expect_error(
      wl_convert_units(1000, "raman", "nm"),
      "Working with Raman shift requires knowledge of laser wavelength"
    )
    expect_error(wl_convert_units(1000, "non-existing", "nm"), "Unknown unit type")
    expect_error(wl_convert_units(1000, "nm", "non-existing"), "Unknown unit type")
  })


  test_that("wl_convert_units() output is correct if units do not change", {
    # No conversion is expected
    expect_equal(wl_convert_units(1000, "raman", "raman"), 1000)
    expect_equal(wl_convert_units(1000, "invcm", "invcm"), 1000)
    expect_equal(wl_convert_units(1000, "nm",    "nm"),    1000)
    expect_equal(wl_convert_units(1000, "ev",    "ev"),    1000)
    expect_equal(wl_convert_units(1000, "freq", "freq"),   1000)
  })


  test_that("wl_convert_units() returns correct data type", {

    x <- c("raman", "invcm", "nm", "ev", "freq")
    y <- expand.grid(x, x)
    y <- y[y[[1]] != y[[2]], ]

    expect_silent(
      d <- apply(y, MARGIN = 1, function(x) {
        wl_convert_units(10, x[["Var1"]], x[["Var2"]], 200)
      })
    )
    expect_is(d, "numeric")
  })

  # TODO (tests): Add expected results to the conversion grid and check against them.

  # test_that("wl_convert_units() performs conversion correctly", {
  #  # ...
  #
  # })

}



R/read_spe.R
# Reading of SPE files, produced by Princeton Instruments spectrometers
# File format version 2.5 (Sept. 2002)

# C. Beleites
# R. Kiselev
# July 2015


# Function -------------------------------------------------------------------

#' Import spectra from WinSpec SPE file
#'
#' Import function for WinSpec SPE files (file version up to 3.0).
#' The calibration data (polynomial and calibration data pairs) for x-axis
#' are automatically read and applied to the spectra. **Note** that
#' the y-calibration data structure is not extracted from the file
#' since it is not saved there by WinSpec and is always empty.
#'
#' @name read_spe
#' @rdname read_spe
#'
#'
#' @param file (character): Name of the SPE file to read data from.
#'
#' @param wl_units (character|`NULL`): Units of x-axis, e.g., `"px"`, `"nm"`,
#'        `"energy"`, `"raman"`, `...`. `NULL` means default units defined in
#'        a file.
#'
#'    Function [read_spe()] automatically checks if the x-calibration data are
#' available and uses them (if possible) to reconstruct the *wavelength* axis
#' in the selected units.
#'
#' @param acc2avg (logical): Whether to divide the actual data set by
#'        the number of accumulations, thus transforming *accumulated*
#'        spectra to *averaged* spectra.
#'        WinSpec does not do this automatically, so the  spectral
#'        intensity is always proportional to the number of accumulations.
#'        The flag `@data$averaged` is automatically set to `TRUE`.
#'
#' @param cts_sec (logical): Whether to divide the actual data set by the
#'        exposure time,  thus going to count per second unit.
#'
#' @param keys_hdr2data (character): Which metadata from the file header
#'        should be saved to the `@data` slot of a newly created
#'        `hyperSpec` object.
#'
#' @param xaxis (DEPRECATED): use `wl_units` instead.
#'
#'
#' @return [hyperSpec][hyperSpec::hyperSpec-class()] object.
#'
#'
#' @author R. Kiselev, C. Beleites
#'
#'
#' @tests testthat
#'
#' # File names
#'
#' # polystyrene <- "fileio\\spe\\polystyrene.SPE"
#' f_blut1 <- system.file(
#'   "extdata",
#'   "blut1.SPE",
#'   package = "hySpc.read.spe"
#' )
#'
#' f_spe3 <- system.file(
#'   "extdata",
#'   "spe_format_3.0.SPE",
#'   package = "hySpc.read.spe"
#' )
#'
#' # Unit tests for `read_spe` itself
#' test_that("read_spe correctly extracts spectral data from SPE file", {
#'   fname <- f_blut1
#'   expect_true(file.exists(fname))
#'   spc <- read_spe(fname)
#'
#'   # We check that specific values are correctly read from a particular file
#'   # This test is not generic and works with this and only this SPE file
#'   expect_equal(spc$spc[[5, 77]], 1484)
#'   expect_equal(spc$spc[[2, 811]], 606)
#'   expect_equal(spc@wavelength[621], 2618.027)
#' })
#'
#' test_that("read_spe() wl_units values", {
#'   fname <- f_blut1
#'
#'   expect_silent(spc_default <- read_spe(fname, wl_units = NULL))
#'   expect_silent(spc_px <- read_spe(fname, wl_units = "px"))
#'   expect_silent(spc_nm <- read_spe(fname, wl_units = "nm"))
#'
#'   expect_match(as.character(labels(spc_default)$.wavelength), "Raman")
#'   expect_match(as.character(labels(spc_nm)$.wavelength), "nm")
#'   expect_match(as.character(labels(spc_px)$.wavelength), "pixel")
#' })
#'
#' test_that("read_spe(): arg. 'xaxis' is deprecated. ", {
#'   fname <- f_blut1
#'
#'   expect_warning(spc_default_d <- read_spe(fname, xaxis = NULL), "deprecated")
#'   expect_silent(spc_default_ok <- read_spe(fname, wl_units = NULL))
#'
#'   expect_equal(spc_default_d, spc_default_ok)
#' })
#'
#'
#' test_that("read_spe detects an XML footer in SPE 3.0 file", {
#'   fname <- f_spe3
#'   expect_true(file.exists(fname))
#'   spc <- read_spe(fname)
#'
#'   expect_true("xml" %in% names(spc@data))
#' })
#'
#'
#' test_that(paste(
#'   "read_spe correctly parses XML footer with SPE 3.0 file and",
#'   "saves metadata in hyperSpec object"
#' ), {
#'   fname <- f_spe3
#'   expect_true(file.exists(fname))
#'   spc <- read_spe(fname)
#'
#'   expect_equal(
#'     attr(spc$xml$SpeFormat$DataFormat$DataBlock, "pixelFormat"),
#'     "MonochromeFloating32"
#'   )
#' })
#'
#'
#' # Unit tests for helper functions of `read_spe` (whose name starts with .)
#' test_that("read_spe_xml_string throws error on SPE format below v3.0", {
#'   fname <- f_blut1
#'   expect_true(file.exists(fname))
#'
#'   expect_error(read_spe_xml_string(fname), regexp = "*no XML*")
#' })
#'
#'
#' test_that("We can correctly read XML footer from SPE3.0 file", {
#'   expect_true(file.exists(f_spe3))
#'
#'   xml_file <- paste0(f_spe3, "_metadata.xml")
#'   actual_xml_footer <- read_spe_xml_string(f_spe3)
#'   expected_xml_footer <- readChar(xml_file, file.info(xml_file)$size)
#'   expect_equal(actual_xml_footer, expected_xml_footer)
#' })
#'
#'
#' test_that(paste(
#'   "read_spe_xml correctly parses the XML footer and",
#'   " can extract the actual data"
#' ), {
#'   expect_true(file.exists(f_spe3))
#'
#'   # Read XML footer and convert it to R list
#'   x <- read_spe_xml(f_spe3)
#'   expect_true(is.list(x))
#'
#'   # Check values of some elements
#'   expect_true(is.list(x))
#'   expect_true("SpeFormat" %in% names(x))
#'
#'   # Check file format version and namespace URL
#'   expect_equal(attr(x$SpeFormat, "version"), "3.0")
#'   expect_equal(
#'     attr(x$SpeFormat, "xmlns"),
#'     "http://www.princetoninstruments.com/spe/2009"
#'   )
#'
#'   # Check that some children are present
#'   expect_true("DataFormat" %in% names(x$SpeFormat))
#'   expect_true("Calibrations" %in% names(x$SpeFormat))
#'   expect_true("DataHistories" %in% names(x$SpeFormat))
#'   expect_true("GeneralInformation" %in% names(x$SpeFormat))
#'
#'   # Check that we can correctly extract file creation date
#'   info <- x$SpeFormat$GeneralInformation$FileInformation
#'   expect_equal(attr(info, "created"), "2018-01-26T16:31:09.0979397+01:00")
#'
#'   # Check that we can correctly extract pixel format and laser line
#'   expect_equal(
#'     attr(x$SpeFormat$DataFormat$DataBlock, "pixelFormat"),
#'     "MonochromeFloating32"
#'   )
#'   expect_equal(
#'     attr(x$SpeFormat$Calibrations$WavelengthMapping, "laserLine"),
#'     "785"
#'   )
#' })
#'
#'
#' @importFrom graphics title
#' @importFrom methods new
#'
#' @export
#'
#' @examples
#' suppressPackageStartupMessages({
#'   library(hySpc.read.spe)
#' })
#'
#' # Example 1
#' f_blut1 <- system.file("extdata", "blut1.SPE", package = "hySpc.read.spe")
#' spc_blut1 <- read_spe(f_blut1)
#'
#' spc_blut1
#'
#' plot(spc_blut1)
#'
#'
#' # Example 2
#' f_spe3 <- system.file(
#'   "extdata", "spe_format_3.0.SPE",
#'   package = "hySpc.read.spe"
#' )
#' spc_spe3 <- read_spe(f_spe3)
#'
#' spc_spe3
#'
#' plot(spc_spe3)
#'
read_spe <- function(file, wl_units = NULL, acc2avg = FALSE, cts_sec = FALSE,
                     keys_hdr2data = c(
                       "exposure_sec",
                       "LaserWavelen",
                       "accumulCount",
                       "numFrames",
                       "darkSubtracted"
                     ),
                     xaxis = "DEPRECATED") {

  if (is.null(xaxis) || (xaxis != "DEPRECATED")) {
    warning("Argument 'xaxis' is deprecated. Use 'wl_units' instead. ")
    wl_units <- xaxis
  }


  hdr <- read_spe_header(file)

  # This is the size of one data point in bytes.
  # WinSpec uses 2 bytes or 4 bytes only.
  data_size <- ifelse(hdr$datatype > 2, 2L, 4L)
  data_chunk_size <- hdr$xdim * hdr$ydim * hdr$numFrames * data_size

  # Read the part of file that contains actual experimental data
  raw_data <- readBin(file, "raw", data_chunk_size + 4100, 1)[-(1:4100)]

  # Convert raw spectral data according to the datatype defined in the header
  spc <- switch(hdr$datatype + 1,
    readBin(raw_data, "double", length(raw_data) / 4, 4), # float
    readBin(raw_data, "integer", length(raw_data) / 4, 4, signed = TRUE), # long
    readBin(raw_data, "integer", length(raw_data) / 2, 2, signed = TRUE), # int
    readBin(raw_data, "integer", length(raw_data) / 2, 2, signed = FALSE) # uint
  )

  # Create a structured data.frame that accommodates spectral data
  dim(spc) <- c(hdr$xdim, hdr$ydim * hdr$numFrames)
  extra_data <- data.frame(
    px.y = rep(seq_len(hdr$ydim), hdr$numFrames),
    frame = rep(seq_len(hdr$numFrames), each = hdr$ydim)
  )

  # Extract selected items from the header.
  # They will go to a new hyperSpec object
  hdr2data <- hdr[keys_hdr2data]
  if (length(hdr2data > 0)) {
    extra_data <- cbind(extra_data, hdr2data)
  }

  # Create hyperSpec object
  spc <- new("hyperSpec",
    spc = t(spc),
    data = extra_data,
    labels = list(spc = "counts", .wavelength = "pixel number")
  )

  # For SPE 3.0 and above we need to read the XML header
  if (hdr$fileFormatVer >= 3.0) {
    spc@data$xml <- read_spe_xml(file)
  }

  # Check if we should use display units specified in the SPE file
  if (is.null(wl_units)) {
    wl_units <- .wl_fix_unit_name(hdr$xCalDisplayUnit)
  }

  # Create a new x-axis, if required
  wl_units <- .wl_fix_unit_name(wl_units)
  if (wl_units == "px") {
    return(.spc_io_postprocess_optional(spc, file))
  }


  if (!hdr$xCalValid) {
    warning("The calibration is NOT valid")
  }

  # Recreate calibration function
  polyorder <- hdr$xCalPolyOrder
  coeffs <- hdr$xCalPolCoeffs[seq(polyorder + 1)]

  vM <- vanderMonde(spc@wavelength, polyorder)

  # Check if we have laser wavelength
  if (hdr$LaserWavelen < 10) {
    hdr$LaserWavelen <- NULL
  }

  # Perform conversion
  spc@wavelength <- wl_convert_units(
    from   = .wl_fix_unit_name(hdr$xCalPolyUnit),
    to     = wl_units,
    x      = as.numeric(vM %*% coeffs),
    ref_wl = hdr$LaserWavelen
  )

  spc@label$.wavelength <-
    switch(wl_units,
      nm    = expression("Wavelength, nm"),
      invcm = expression(tilde(nu) / cm^-1),
      ev    = expression("Energy / eV"),
      freq  = expression(nu / THz),
      raman = expression(Raman ~ shift / cm^-1)
    )
  if (acc2avg) {
    spc <- spc / hdr$accumulCount
    spc@data$averaged <- TRUE
  }
  if (cts_sec) {
    spc <- spc / hdr$exposure_sec
    spc@label$spc <- expression("counts / s")
  }

  ## consistent file import behavior across import functions
  .spc_io_postprocess_optional(spc, file)
}


# Helper functions -----------------------------------------------------------

#' Read XML footer from SPE file format version 3.0
#'
#' The new SPE file format, introduced in 2012, was designed to be backwards
#' compatible with the previous format 2.5. The most prominent change is the
#' new plain text XML footer holding vast experimental metadata that gets
#' attached at the end of the file. Thus, the file contains 3 blocks:
#' a 4100-bytes long binary header, a chunk with spectral data, and the XML
#' footer.
#' This function retrieves the XML footer converted to R list, and throws error
#' if it is not available.
#' The file format specification is available at Princeton Instruments FTP
#' server under name 'SPE 3.0 File Format Specification'.
#'
#' This function relies on R package \pkg{xml2} to work correctly.
#'
#' @param file Path to SPE file.
#'
#' @return XML data from the file converted to R list.
#'
#' @importFrom xml2 as_list read_xml
#'
#' @noRd
read_spe_xml <- function(file) {
  as_list(read_xml(read_spe_xml_string(file)))
}


#' read_spe_xml_string
#'
#' Read XML footer from SPE file format version 3.0 and return it as a long
#' string for subsequent parsing. Basically the purpose of this function is
#' to check that the file format version is 3.0 or above, and to find and
#' read the correct part of this file.
#'
#' @param file Path to SPE file.
#'
#' @return string containing XML footer
#'
#' @noRd

read_spe_xml_string <- function(file) {
  hdr <- read_spe_header(file)

  if (hdr$fileFormatVer < 3.0) {
    stop(paste(
      "This SPE file contains no XML data: file format version",
      round(hdr$fileFormatVer, digits = 3), "< 3.0"
    ))
    return()
  }

  data_size <- ifelse(hdr$datatype > 2, 2L, 4L)
  data_chunk_size <- hdr$xdim * hdr$ydim * hdr$numFrames * data_size

  # Read the part of file that contains actual experimental data
  i_data <- -(1:(4100 + data_chunk_size))
  raw_bytes <- readBin(file, "raw", file.info(file)$size, 1)[i_data]
  readChar(raw_bytes, length(raw_bytes))
}


# @describeIn read_spe
#
#' Read only header of a WinSpec SPE file (version 2.5)
#'
#' @return hdr list with `key=value` pairs
#'
#' @noRd

read_spe_header <- function(file) {
  # Read the 4100-byte long binary header from the SPE file and parse it

  # Load the header
  raw_data <- readBin(file, "raw", 4100, 1)

  # Extract some items from the 4100 bytes-long file header
  hdr <- list(
    hwVersion       = readBin(raw_data[1:2], "integer", 1, 2, signed = TRUE), # uint16
    xDimDet         = readBin(raw_data[7:8], "integer", 1, 2, signed = FALSE), # uint16
    mode            = readBin(raw_data[9:10], "integer", 1, 2, signed = TRUE), # uint16
    exposure_sec    = readBin(raw_data[11:14], "double", 1, 4), # float32
    vChipXDim       = readBin(raw_data[15:16], "integer", 1, 2, signed = TRUE), # int8
    vChipYDim       = readBin(raw_data[17:18], "integer", 1, 2, signed = TRUE), # int8
    yDimDet         = readBin(raw_data[19:20], "integer", 1, 2, signed = FALSE), # uint16
    date            = readBin(raw_data[21:30], "character", 1, 10), # char
    detTemperature  = readBin(raw_data[37:40], "double", 1, 4), # float32
    xdim            = readBin(raw_data[43:44], "integer", 1, 2, signed = FALSE), # uint16
    shutterMode     = readBin(raw_data[51:52], "integer", 1, 2, signed = FALSE), # uint16
    specCenterWlNm  = readBin(raw_data[73:76], "double", 1, 4), # float32
    datatype        = readBin(raw_data[109:110], "integer", 1, 2, signed = TRUE), # int8
    darkSubtracted  = readBin(raw_data[151:152], "integer", 1, 2, signed = FALSE), # int8
    timeLocal       = readBin(raw_data[173:179], "character", 1, 7), # char
    timeUTC         = readBin(raw_data[180:186], "character", 1, 7), # char
    gain            = readBin(raw_data[199:200], "integer", 1, 2, signed = FALSE), # uint16
    comments        = readBin(raw_data[201:600], "character", 1, 400), # char
    ydim            = readBin(raw_data[657:658], "integer", 1, 2, signed = FALSE), # uint16
    accumulCount    = readBin(raw_data[669:672], "integer", 1, 4), # uint32
    readoutTime     = readBin(raw_data[673:676], "double", 1, 4), # float32
    swVersion       = readBin(raw_data[688:704], "character", 1, 16), # char
    kinTrigMode     = readBin(raw_data[725:726], "integer", 1, 2, signed = TRUE), # int16
    expRepeatCount  = readBin(raw_data[1419:1422], "integer", 1, 4, signed = TRUE), # int32
    expAccumCount   = readBin(raw_data[1423:1426], "integer", 1, 4, signed = TRUE), # int32
    hwAccumFlag     = readBin(raw_data[1433:1434], "integer", 1, 2, signed = TRUE), # int16
    cosmicApplied   = readBin(raw_data[1439:1440], "integer", 1, 2, signed = TRUE), # int16
    cosmicType      = readBin(raw_data[1441:1442], "integer", 1, 2, signed = TRUE), # int16
    numFrames       = readBin(raw_data[1447:1450], "integer", 1, 4), # int32
    shutterType     = readBin(raw_data[1475:1476], "integer", 1, 2, signed = TRUE), # int16
    readoutMode     = readBin(raw_data[1481:1482], "integer", 1, 2, signed = TRUE), # int16
    kinWindowSize   = readBin(raw_data[1483:1484], "integer", 1, 2, signed = TRUE), # int16
    clkSpeed        = readBin(raw_data[1485:1486], "integer", 1, 2, signed = TRUE), # int16
    computerIface   = readBin(raw_data[1487:1488], "integer", 1, 2, signed = TRUE), # int16
    fileFormatVer   = readBin(raw_data[1993:1996], "double", 1, 4, signed = TRUE), # float32

    # X Calibration Structure
    xCalOffset      = readBin(raw_data[3001:3008], "double", 1, 8, signed = TRUE), # float64
    xCalFactor      = readBin(raw_data[3009:3016], "double", 1, 8, signed = TRUE), # float64
    xCalDisplayUnit = readBin(raw_data[3017], "integer", 1, 1, signed = FALSE), # uint8
    xCalValid       = readBin(raw_data[3099], "integer", 1, 1, signed = FALSE), # uint8
    xCalInputUnit   = readBin(raw_data[3100], "integer", 1, 1, signed = FALSE), # uint8
    xCalPolyUnit    = readBin(raw_data[3101], "integer", 1, 1, signed = FALSE), # uint8
    xCalPolyOrder   = readBin(raw_data[3102], "integer", 1, 1, signed = FALSE), # uint8
    xCalPointCount  = readBin(raw_data[3103], "integer", 1, 1, signed = FALSE), # uint8
    xCalPxPos       = readBin(raw_data[3104:3183], "double", 10, 8, signed = TRUE), # float64
    xCalValues      = readBin(raw_data[3184:3263], "double", 10, 8, signed = TRUE), # float64
    xCalPolCoeffs   = readBin(raw_data[3264:3311], "double", 6, 8, signed = TRUE), # float64
    LaserWavelen    = readBin(raw_data[3312:3319], "double", 1, 8, signed = TRUE) # float64
  )

  # Convert magic numbers into human-readable unit strings
  spe_units <- c("pixel", "pixel", "data", "user units", "nm", "cm-1", "Raman shift")
  hdr$xCalDisplayUnit <- spe_units[hdr$xCalDisplayUnit + 1]
  hdr$xCalInputUnit <- spe_units[hdr$xCalInputUnit + 1]
  hdr$xCalPolyUnit <- spe_units[hdr$xCalPolyUnit + 1]

  return(hdr)
}
